src/main/java/games/strategy/triplea/oddsCalculator/ta/ConcurrentOddsCalculator.java

```diff
@@ -1,437 +1,58 @@
-      // use half the number of threads available if we took
-      // more than 3 seconds to copy
-      return Math.min(numberOfTimesWeCanCopyMax, Math.max(1, (MAX_THREADS / 2)));
-    }
-    // use all threads
-    return Math.min(numberOfTimesWeCanCopyMax, MAX_THREADS);
-  }
-
-  private void createWorkers(final GameData data) {
-    workers.clear();
-    if (data != null && cancelCurrentOperation >= 0) {
-      // see how long 1 copy takes (some games can get REALLY big)
-      final long startTime = System.currentTimeMillis();
-      final long startMemory = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
-      final GameData newData;
-      try {
-        // make first copy, then release lock on it so game can continue (ie: we don't want to lock on it while we copy
-        // it 16 times, when once is enough) don't let the data change while we make the first copy
-        data.acquireReadLock();
-        newData = GameDataUtils.cloneGameData(data, false);
-      } finally {
-        data.releaseReadLock();
-      }
-      currentThreads = getThreadsToUse((System.currentTimeMillis() - startTime), startMemory);
-      try {
-        // make sure all workers are using the same data
-        newData.acquireReadLock();
-        int i = 0;
-        // we are already in 1 executor thread, so we have MAX_THREADS-1 threads left to use
-        if (currentThreads <= 2 || MAX_THREADS <= 2) {
-          // if 2 or fewer threads, do not multi-thread the copying (we have already copied it once above, so at most
-          // only 1 more copy to
-          // make)
-          while (cancelCurrentOperation >= 0 && i < currentThreads) {
-            // the last one will use our already copied data from above, without copying it again
-            workers.add(new OddsCalculator(newData, (currentThreads == ++i)));
-          }
-        } else { // multi-thread our copying, cus why the heck not (it increases the speed of copying by about double)
-          final CountDownLatch workerLatch = new CountDownLatch(currentThreads - 1);
-          while (i < (currentThreads - 1)) {
-            ++i;
-            executor.submit(() -> {
-              if (cancelCurrentOperation >= 0) {
-                workers.add(new OddsCalculator(newData, false));
-              }
-              workerLatch.countDown();
-            });
-          }
-          // the last one will use our already copied data from above, without copying it again
-          workers.add(new OddsCalculator(newData, true));
-          try {
-            workerLatch.await();
-          } catch (final InterruptedException e) {
-            Thread.currentThread().interrupt();
-          }
+    final int totalRunCount = parameters.runCount;
+    parameters.setRunCount(1);
+    for (int i = 0; i < totalRunCount; i++) {
+      executor.submit(() -> {
+        if (!cancelled) {
+          aggregateResults.addResult(OddsCalculator.doSimulation(parameters));
```

Comment:
Multiple threads are calling the `AggregateResults#addResult()` method, but access to the corresponding `AggregateResults#m_results` field does not appear to be thread safe (there are no locks and the field is of the thread-unsafe type `ArrayList`).

Reply:
4960d79